File: samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/infrastructure/zlink/spots/entryspot/handlers/MatchBingoActorHandler.java
package systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.entryspot.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.spots.ZLinkEntrySpotActorRequestHandler;
import systems.zlink.samples.bingo.server.configuration.SampleNames;
import systems.zlink.samples.bingo.server.configuration.SampleTimings;
import systems.zlink.samples.bingo.server.play.infrastructure.zlink.actors.PlayerActor;
import systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.entryspot.BingoEntrySpot;
import systems.zlink.samples.bingo.shared.contracts.BingoMessages;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.util.List;
import java.util.concurrent.CompletionStage;

public final class MatchBingoActorHandler
        implements ZLinkEntrySpotActorRequestHandler<
                BingoEntrySpot, PlayerActor, Messages.MatchBingoReq, Messages.MatchBingoRes> {
    @Override
    public CompletionStage<Messages.MatchBingoRes> handle(
            BingoEntrySpot entrySpot,
            PlayerActor actor,
            ZLinkMessageContext context,
            Messages.MatchBingoReq request) {
        // --8<-- [start:doc-bingo-match-actor]
        return entrySpot
                .context()
                .outbound()
                .requestToChannel(
                        SampleNames.ApiChannel,
                        BingoMessages.matchBingoApiReq(
                                actor.actorId(), actor.displayName(), request.getMode()))
                .timeout(SampleTimings.RequestTimeout)
                .submit(Messages.MatchBingoApiRes.class)
                .thenApply(
                        matched -> {
                            actor.trackDeferredJoin(matched.getRoomId());
                            actor.context()
                                    .joinSpot(
                                            matched.getRoomId(),
                                            BingoMessages.bingoRoomJoinReq(
                                                    matched.getRoomId(),
                                                    actor.actorId(),
                                                    actor.displayName(),
                                                    false))
                                    .timeout(SampleTimings.RequestTimeout)
                                    .defer();
                            Messages.BingoRoomState initialState =
                                    BingoMessages.bingoRoomState(
                                            matched.getRoomId(),
                                            "WaitingForPlayers",
                                            actor.actorId(),
                                            false,
                                            0,
                                            null,
                                            List.of(),
                                            List.of(
                                                    BingoMessages.bingoPlayerState(
                                                            actor.actorId(),
                                                            actor.displayName(),
                                                            1,
                                                            true,
                                                            List.of(),
                                                            List.of(),
                                                            0,
                                                            0,
                                                            0)),
                                            List.of());
                            return BingoMessages.matchBingoRes(matched.getRoomId(), initialState);
                        });
        // --8<-- [end:doc-bingo-match-actor]
    }
}

File: samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/infrastructure/zlink/spots/entryspot/handlers/MatchBingoActorHandler.kt
package systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.entryspot.handlers

import kotlinx.coroutines.future.await
import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingEntrySpotActorRequestHandler
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleNames
import systems.zlink.samples.kotlin.bingo.server.configuration.SampleTimings
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.actors.PlayerActor
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.entryspot.BingoEntrySpot
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoPlayerState
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomJoinReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.BingoRoomState
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoApiReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoApiRes
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.MatchBingoRes

class MatchBingoActorHandler :
    ZLinkSuspendingEntrySpotActorRequestHandler<
        BingoEntrySpot,
        PlayerActor,
        MatchBingoReq,
        MatchBingoRes,
    > {
    override suspend fun handle(
        entrySpot: BingoEntrySpot,
        actor: PlayerActor,
        context: ZLinkMessageContext,
        message: MatchBingoReq,
    ): MatchBingoRes {
        // --8<-- [start:doc-bingo-match-actor]
        val matched =
            entrySpot
                .context()
                .outbound()
                .requestToChannel(
                    SampleNames.ApiChannel,
                    MatchBingoApiReq(actor.actorId(), actor.displayName, message.mode),
                )
                .timeout(SampleTimings.RequestTimeout)
                .submit(MatchBingoApiRes::class.java)
                .await()
        actor.trackDeferredJoin(matched.roomId)
        actor
            .context()
            .joinSpot(
                matched.roomId,
                BingoRoomJoinReq(matched.roomId, actor.actorId(), actor.displayName, false),
            )
            .timeout(SampleTimings.RequestTimeout)
            .defer()
        // --8<-- [end:doc-bingo-match-actor]
        return MatchBingoRes(
            matched.roomId,
            BingoRoomState(
                roomId = matched.roomId,
                status = "WaitingForPlayers",
                hostActorId = actor.actorId(),
                canStart = false,
                drawSeq = 0,
                lastDrawnNumber = null,
                drawnNumbers = emptyList(),
                players =
                    listOf(
                        BingoPlayerState(
                            actorId = actor.actorId(),
                            displayName = actor.displayName,
                            seat = 1,
                            isHost = true,
                            card = emptyList(),
                            marks = emptyList(),
                            completedLines = 0,
                            wins = 0,
                            losses = 0,
                        )
                    ),
                winners = emptyList(),
            ),
        )
    }
}
